### PR TITLE
add fashion ai

### PR DESCRIPTION
**Pull request by sadiea**

### CODE REVIEW

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+import websockets
+import threading
+from datetime import datetime
+import time
+
+
+APP_ID = ""
+CHANNEL = ""
+STREAM_ID = "0"
```
Comment (plutoless) [inline agents/ten_packages/extension/fashionai/src/extension.py:32]: maybe add as instance properties

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    async def init_fashionai(self, ten_env: TenEnv, app_id, channel, stream_id):
+        await self.client.connect()
+        await self.client.stream_start(app_id, channel, stream_id)
+        # asyncio.create_task(self.client.heartbeat(10))
+        await self.client.render_start()
+        while True:
+            await self.async_polly_handler()
+
+    def thread_target(self, loop, ten_env: TenEnv, app_id, channel, stream_id):
```
Comment (plutoless) [inline agents/ten_packages/extension/fashionai/src/extension.py:130]: maybe not necessary to make it instance method, just make it a local func in on_start?

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            except Exception as e:
+                logger.exception(e)
+
+    def flush(self):
+        logger.info("FASHION_AI flush")
+        while not self.queue.empty():
+            self.queue.get()
+
+
+class WebSocketClient:
```
Comment (plutoless) [inline agents/ten_packages/extension/fashionai/src/extension.py:152]: maybe call it FashionAIClient

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                logger.info(f"FASHION_AI Received: {response}")
+            except websockets.exceptions.ConnectionClosedError as e:
+                logger.info(f"FASHION_AI Connection closed with error: {e}")
+                await self.reconnect()
+            except asyncio.TimeoutError:
+                logger.info("FASHION_AI Timeout waiting for response")
+        else:
+            logger.info("FASHION_AI WebSocket is not connected.")
+
+    async def receive_message(self):
```
Comment (plutoless) [inline agents/ten_packages/extension/fashionai/src/extension.py:208]: not used?

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        if len(TOKEN) > 0 :
+            APP_ID = TOKEN
+        logger.info(f"FASHION_AI on_start: app_id = {APP_ID}, token = {TOKEN}, channel = {CHANNEL}, stream_id = {STREAM_ID}")
+
+        new_loop = asyncio.new_event_loop()
+
+        self.threadWebsocket = threading.Thread(target=self.thread_target, args=(new_loop,ten_env,APP_ID ,CHANNEL,STREAM_ID))
+        self.threadWebsocket.start()
+
+        threading.Thread(target=self.async_polly_handler, args=[ten_env]).start()
```
Comment (plutoless) [inline agents/ten_packages/extension/fashionai/src/extension.py:63]: not needed?

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+    def on_data(self, ten_env: TenEnv, data: Data) -> None:
+        # TODO: process data
+        inputText = data.get_property_string("text")
+        if len(inputText) == 0:
+            logger.info("FASHION_AI ignore empty text")
+            return
+
+        logger.info("FASHION_AI on data %s", inputText)
+        asyncio.run(self.queue.put((inputText, datetime.now())))
```
Comment (plutoless) [inline agents/ten_packages/extension/fashionai/src/extension.py:105]: this will lead to new thread. use run_coroutine_threadsafe or put_nowait

Comment (plutoless) [reply agents/ten_packages/extension/fashionai/src/extension.py:105]: asyncio.run is not supposed to be called multiple times.

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    def on_video_frame(self, ten_env: TenEnv, video_frame: VideoFrame) -> None:
+        # TODO: process image frame
+        pass            
+
+    async def init_fashionai(self, ten_env: TenEnv, app_id, channel, stream_id):
+        await self.client.connect()
+        await self.client.stream_start(app_id, channel, stream_id)
+        # asyncio.create_task(self.client.heartbeat(10))
+        await self.client.render_start()
+        while True:
```
Comment (plutoless) [inline agents/ten_packages/extension/fashionai/src/extension.py:126]: maybe put wihle true in async_polly_handler

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+import websockets
+import threading
+from datetime import datetime
+
+class FashionAIExtension(Extension):
+
+
+    def on_init(self, ten_env: TenEnv) -> None:
+        logger.info("FASHION_AI on_init *********************************************************")
+        self.app_id = ""
```
Comment (tomasliu-agora) [inline agents/ten_packages/extension/fashionai/src/extension.py:33]: why not define them in class directly...

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+        self.loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(self.loop)
+
+        ten_env.on_init_done()
+
+    def on_start(self, ten_env: TenEnv) -> None:
+        logger.info("FASHION_AI on_start *********************************************************")
+
+        # TODO: read properties, initialize resources
+        self.app_id = ten_env.get_property_string("app_id")
```
Comment (tomasliu-agora) [inline agents/ten_packages/extension/fashionai/src/extension.py:49]: you need try catch here.

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    def on_cmd(self, ten_env: TenEnv, cmd: Cmd) -> None:
+        cmd_name = cmd.get_name()
+        logger.info("FASHION_AI on_cmd name {}".format(cmd_name))
+
+        # TODO: process cmd
+        if cmd_name == "flush":
+            self.outdate_ts = datetime.now()
+            asyncio.run_coroutine_threadsafe(
+                self.flush(), self.loop
+            ).result(timeout=0.1)
```
Comment (tomasliu-agora) [inline agents/ten_packages/extension/fashionai/src/extension.py:89]: what will happen when timeout?

File: agents/ten_packages/extension/fashionai/src/extension.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+                    logger.exception(e)
+
+    async def flush(self):
+        logger.info("FASHION_AI flush")
+        while not self.queue.empty():
+            value = await self.queue.get()
+            logger.info(f"Flushing value: {value}")
+
+
+class FashionAIClient:
```
Comment (tomasliu-agora) [inline agents/ten_packages/extension/fashionai/src/extension.py:152]: should this be a separated file?

**Pull request closed without merging**